File: src/main/java/com/pedigree/editor/commands/AddFamilyCommand.java
package com.pedigree.editor.commands;

import com.pedigree.editor.Command;
import com.pedigree.model.Family;
import com.pedigree.storage.ProjectRepository;

import java.util.Objects;

public class AddFamilyCommand implements Command {
    private final ProjectRepository.ProjectData data;
    private final Family familyToAdd;
    private boolean added;

    public AddFamilyCommand(ProjectRepository.ProjectData data, Family familyToAdd) {
        this.data = Objects.requireNonNull(data, "data");
        this.familyToAdd = Objects.requireNonNull(familyToAdd, "familyToAdd");
    }

    @Override
    public void execute() {
        boolean exists = data.families.stream().anyMatch(f -> f.getId().equals(familyToAdd.getId()));
        if (exists) {
            throw new IllegalStateException("Family with id already exists: " + familyToAdd.getId());
        }
        data.families.add(familyToAdd);
        added = true;
    }

    @Override
    public void undo() {
        if (!added) { return; }
        data.families.removeIf(f -> f.getId().equals(familyToAdd.getId()));
        added = false;
    }
}

File: src/main/kotlin/com/family/tree/editor/commands/AddFamilyCommand.kt
package com.family.tree.editor.commands

import com.family.tree.editor.Command
import com.family.tree.model.Family
import com.family.tree.storage.ProjectRepository
import java.util.Objects

class AddFamilyCommand(
    private val data: ProjectRepository.ProjectData,
    private val familyToAdd: Family
) : Command {
    private var added: Boolean = false

    init {
        Objects.requireNonNull(data, "data")
        Objects.requireNonNull(familyToAdd, "familyToAdd")
    }

    override fun execute() {
        val exists = data.families.any { it.id == familyToAdd.id }
        if (exists) {
            throw IllegalStateException("Family with id already exists: ${familyToAdd.id}")
        }
        data.families.add(familyToAdd)
        added = true
    }

    override fun undo() {
        if (!added) return
        data.families.removeIf { it.id == familyToAdd.id }
        added = false
    }
}
